### ISSUE TITLE
Implementing a custom IServer is not as easy as it should be now that some classes have been made internal

### ISSUE THREAD

**Issue by roblappcoupa** [dotnet/aspnetcore#38503]
#### Context
I am hoping to re-examine [this feature request](https://github.com/dotnet/aspnetcore/issues/33913). For learning purposes, I wanted to create a custom `IServer` that leverages the ASP.NET Core request pipeline. Previously, this was fairly straight forward. An example of this is [here](https://reynders.co/use-iserver-from-aspnet-core-to-create-your-own-web-server/). In the example the request pipeline is invoked by using a `FileSystemWatcher` instead of a TCP socket. This solution relies on the `HostingApplication.Context` class being accessible in the `IServer` implementation so that you can populate an `HttpContext` and then invoke the request pipeline. This was really nice because it was one extensibility point and everything just seemed to magically work.

#### Problem
The problem is now that `HostingApplication.Context` has been made internal, there is no easy way that I can see to set an `HttpContext` from an `IServer` implementation. I saw that one of the ASP.NET Core developers said to provide an `IHttpContextFactory` implementation. This works nicely if you want to provide a standard `HttpContext`. But what if you want to provide a custom `HttpContext`? Or what if you wanted to trigger the request pipeline from something other than a standard TCP network socket, such as an AMQP message? When `HostingApplication.Context` was accessible, a consumer had full control over 1) how to construct an `HttpContext` and 2) when and how to invoke the request pipeline with minimal effort. While I understand these might be outlier use cases, it would be really nice if this was as straight forward as it once was.

### Describe the solution you'd like
As far as I can tell, to accomplish something like an AMQP driven request pipeline or the example that uses a `FileSystemWatcher`, you'd end up having to override and re-implement or duplicate a lot of core functionality that exists in the Hosting and Servers projects. Given that many classes are `internal`, I am not sure how possible it is. A simple and working solution would be to have the `Context` that is given to the `IServer` implementation be accessible so that it can be set by the custom `IServer` and then used in the subsequent invocation of `application.ProcessRequestAsync`. For example, consider a (ridiculous) console driven `IServer` implementation that is intended only to demonstrate a possible solution:

```
public Task StartAsync<TContext>(IHttpApplication<TContext> application, CancellationToken cancellationToken)
        where TContext : notnull
    {
        return Task.Run(
            async () =>
            {
                while (!cancellationToken.IsCancellationRequested)
                {
                    try
                    {
                        // Assume the syntax coming from the console was VERB PATH. Example: GET /weather
                        var line = Console.ReadLine();

                        if (string.IsNullOrWhiteSpace(line))
                        {
                            return;
                        }

                         var context = (HostingApplication.Context)(object)application.CreateContext(Features);
                        context.HttpContext = new ConsoleHttpContext(Features, line);

                        await application.ProcessRequestAsync((TContext)(object)context);

                        context.HttpContext.Response.OnCompleted(null, null);
                    }
                    catch (Exception exception)
                    {
                        Console.WriteLine(exception);
                    }
                }
            },
            cancellationToken);
    }
```
The benefit of this is:
1. We can set the properties of the `HttpContext` or a custom `HttpContext` however we want. In this case the data is coming from the console.
2. We can choose when to invoke the request pipeline. In this case, we invoke the request pipeline when the user inputs data.

In the same spirit, if we wanted an AMQP driven request pipeline, the code might look like this:
```
public async Task StartAsync<TContext>(IHttpApplication<TContext> application, CancellationToken cancellationToken)
        where TContext : notnull
    {
        // Note: bus is an IBus instance from the EasyNetQ library
        var result = await bus.PubSub.SubscribeAsync<AmqpRequest>(
            "my_subscription_id",
            async message =>
            {
                var context = (HostingApplication.Context)(object)application.CreateContext(Features);
                context.HttpContext = new AmqpInitiatedHttpContext(Features, line);

                await application.ProcessRequestAsync((TContext)(object)context);

                context.HttpContext.Response.OnCompleted(null, null);
            });
    }
```
Again, this is to demonstrate:
1. We can set the properties of the `HttpContext` or a custom `HttpContext` however we want. In this case the data is coming from an AMQP message.
2. We can choose when to invoke the request pipeline. In this case, we invoke the request pipeline when the application handles an AMQP message.

### Final Thoughts
In summary, it would be nice if it was easy as before to set the `HttpContext` and invoke the request pipeline from a custom `IServer` implementation,.

**Comment by Tratcher** [comment]
I think a key design point you're missing here is HttpContext's intended extensibility pattern. HttpContext is intended to be implemented through the IFeatureCollection, not derivation. The server creates a feature collection per request, fills it with the needed features, and then passes that to the IHttpContextFactory to create the actual HttpContext. Most features are optional, and the required ones like the request and response features have default implementation available here:
https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http/src/Features/HttpRequestFeature.cs

Note the request IFeatureCollection here is not the same instance as IServer.Features.

**Comment by roblappcoupa** [comment]
@Tratcher thanks for the info. Are you saying that it would be better to try and implement the necessary feature interface(s) instead of IServer itself?

**Comment by Tratcher** [comment]
You implement IServer and the feature interfaces instead of deriving from HttpContext.

**Comment by roblappcoupa** [comment]
We can close this out as my use cases are likely outliers. I was just trying to experiment around with the extensibility of the framework as demonstrated in the code samples above as well as the now outdated `AwesomeServer` article. With that being said, from a consumer standpoint, this does not seem as easy as it once was, at least given my current level of knowledge with the code base. I probably need to spend more time studying the code base to understand it better or move forward with something else. Thanks.
